### ISSUE TITLE
Multiple Classes

### ISSUE THREAD

**Issue by Cerebralism** [WyattAutomation/Train-YOLOv3-with-OpenImagesV4#1]
Thank you for creating this repo. It helped me train my own YOLO model.

It says in the read me that you would be updating this to work with multiple classes. Did you get this working? I would really like to see this!

**Comment by WyattAutomation** [comment]
Im going to work on this tonight; sorry for the delay!  It should be up tonight or tomorrow night.

**Comment by cyrineee** [comment]
Thanks a lot for sharing :) .Are you still working for Multiple Classes?

**Comment by WyattAutomation** [comment]
I'm setting a reminder now to get this done.  Work priorities have taken up much of my spare time recently, but I should have a moment to whip this up tonight.  I'll do my best to not leave you folks hanging.

**Comment by cyrineee** [comment]
Can i test ur code with NVIDIA GeForce GT 635 M ? i want to test it first for one class and after for multiclass .

**Comment by WyattAutomation** [comment]
Oof.. I spent all weekend studying and practicing code and didn't get back to this...  Sorry about that; life and things..

Anyway;  in lieu of flaking on you again here, let's at least see if your card is capable of running this thing shall we?

2 crucial things we will need to look at are the amount of vram that card has and the compute architecture.  

I was able to get PJReddie's original YOLOv3 working with a live feed from a webcam at about 18 fps on a legacy Quadro 4000, with 2gb of vram and with 2.0 compute architecture, on xubuntu 16.04.  That was before I saved up for  a better card (1060, 6gb for my poor-mans ML rig).  Getting it working on that Quadro in the meantime was a *total* pain to figure out; 2.0 and 2.1 (2.1 being the mobile "2" arch) architecture was deprecated in cuda 8.0, and at the time, PJ's sparce documentation only mentioned using cuda 9 (probably still does) and in the makefile he even had the compute_20 and 21 arch compatibility settings commented out saying some vague hint like "this is deprecated?.."

..yeah real helpful comment to us plebs that can't afford a new pallet of NVidia 2080s every month you brilliant, incredible person.. I'm gonna go eat some ramen and cry now..

...anyway..

A Google search on that card looks like it may have 1 or 2 gb of vram; looks like there may have been 2 options for that card for memory?  IDK, but in regards to that you'll probably need at least 2.  I doubt that it will work at all with less, and I had to tweak the config heavily to get it at 18fps with 2gb of vram, but I can confirm it is doable.

Let's look up your GPU's compute architecture:

 https://developer.nvidia.com/cuda-gpus

Yuck--you're at compute architecture 2.1 ("sm_21" and "compute_21" iirc), meaning the exact same boat I was in when I was stuck figuring out how to get my 2 gigs of decade old vram to run the thing.

YOLOv3 does work with cuda 8.0, but it is not a fun process getting all the newer-than-cuda8 dependencies that it does require working with it.  

Not to mention are we talking about training with that card.. like in a laptop?  Inference (just running the model after its trained) is one thing but it will be another challenge to dial the config down enough to not have an out of memory error or heat issues.  I trained the one in my tutorial here on that 1060 in a big metal box with lots of fans, and even that took like 10 hours (though I intentionally underutilized it so it would run slow, cool and stable).

Anyway you can probably get it working in terms of at least running inference with pretrained weights on that card but it's going to be a challenge.  I can help if you're up for learning a thing or thirty about cuda libraries, Debian, nvcc, and all sorts of other issues I had to deal with getting legacy arch working with this, but it certainly can be done.

Maybe after I finally get around to updating the data prep script I can make a document on getting yolo working well on inexpensive legacy cards?  That'd be pretty fun tbh.

TLDR:. Install cuda 8 for your rig, adjust the YOLOv3 makefile to allow the _21 architecture, and then try to get it working from there per the 'make' instructions on the main site for yolo .  There will be dozens of issues, more than I have time to unpack in a single post here but it is doable I promise...

**Comment by WyattAutomation** [comment]
Getting back to the original issue; I will get this done I promise.  It isn't so much the script as much as it is the documentation that will take a while to update, and I would rather have matching documentation to the code and have it working rather than just slap it on there and have issues pop up with the instructions no longer working. I may just include a second set of instructions that tell you what you need to do differently, along with the scripts.

Anyway, I will return to this soon and get this issue closed out. Today and tonight I am pretty booked but I will do what I can.

**Comment by WyattAutomation** [comment]
All,
I have added a script here which shows how to process the classes for Beer, Cat and Banana at the same time and can be edited easily to process any number of classes from Open Images:

https://github.com/WyattAutomation/Train-YOLOv3-with-OpenImagesV4/blob/master/MULTI-CLASSjpg2txt.py

Please note that everything in the original README document that was set up in regards to the config files has to be adjusted too in order to account for however many classes you put in here.  

It should be very straight forward, so if you have any questions or issues with this script not working please let me know and I will update the README doc to include more instructions if needed.

**Comment by cyrineee** [comment]
for the training using darknet , any modification apart (.cfg , .names ,.. ) ?

**Comment by WyattAutomation** [comment]
Apart from the configuration instructions that are in the original README document, there shouldn't be anything that doesn't just include adding extra classes per the conventions that are already outlined.  If you have a specific question with a problem you run into while using the new script, open a new issue.

A G-Force 635 M is not practical to use to train this model.. It "can" be done but getting this running on legacy hardware  is several orders of magnitude more difficult to figure out than what I've outlined in this tutorial..

Get it installed first and run the demo with the pretrained demo weights with your webcam using that card, then if you run into a problem with the script not working after installing yolo open a new issue.

If you don't have the resources for a better gpu, I'm willing to help you figure out how to get it working, but I'll need specifics on what you've tried and where you're stuck.  Dive in and give it a shot.

**Comment by Cerebralism** [comment]
@WyattAutomation thanks for doing this. With multiple classes, where do I move the files output by MULTI-CLASSjpg2txt.py ? With a single class you move them to OIDv4_ToolKit/OID/Dataset/test/single_class_name but what should I do when there are multiple class directories?

Same question for process.py file. What do I set current_dir on line 5 do now that there are multiple training class directories?

Thank you again for helping with this :)

**Comment by WyattAutomation** [comment]
You should be able to dump all of the photos for all of the classes you've downloaded into the same directory, then run the multi-class script from that one directory that contains all of the photos.  Afterwards point the process.py at that directory too.   

Basically, just do all the work in the same directory, start to finish.

There isn't a check for duplicates, as some photos may contain both of the classes you want, and this is an issue that I can resolve in an update.  Also, it would be wise to try to either pick classes with roughly the same number of photos in each downloaded class, or adjust it so the same number of photos is used for each class.  I ran into issues where it detected "beer" way better than cat or banana as there were way more photos of beer than the other two.

**Comment by Cerebralism** [comment]
Thank you for the quick response, I'll try this and see what happens. I'm trying this with quite imbalanced classes. I'll leave this as it is for now but maybe try again a second time with balanced numbers. Do you have a recommendation on how best to remove images from the overly-populated classes?

**Comment by WyattAutomation** [comment]
In Linux, you can move the first however-many files in a directory to another directory like so:

`mv 'ls | head -500' ./subfolder1`

Change 500 to whatever number of photos you need to remove.

May need to run that as sudo..  

Be sure to make a backup of the fresh datasets you download too; don't want to screw something up and have to redownload.
